### PR TITLE
Fixes #95 Add EPEL GPG Key and logic to handle yum::gpgkeys

### LINKED ISSUE #95: EPEL GPG Key
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: puppet-agent-5.3.5-1.el7.x86_64
- Ruby: ruby 2.4.3p205 (2017-12-14 revision 61247) [x86_64-linux]
- Distribution: CentOS Linux release 7.4.1708 (Core)
- Module version: 2.2.1

## How to reproduce (e.g Puppet code you use)

```puppet
include ::yum

package { 'rootsh':
  ensure => installed,
}
```

```yaml
yum::managed_repos:
  - epel
```

## What are you seeing

```
Error: /Stage[main]/Profile::Davita_base::Rootsh/Package[rootsh]/ensure: change from 'purged' to 'present' failed: Execution of '/bin/yum -d 0 -e 0 -y install rootsh' returned 1: warning: /var/cache/yum/x86_64/7/epel/packages/rootsh-1.5.3-17.el7.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID 352c64e5: NOKEY


GPG key retrieval failed: [Errno 14] curl#37 - "Couldn't open file /etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7"
```

## What behaviour did you expect instead

I expected the package to be installed ;)

## Output log

see error above

## Any additional information you'd like to impart

The GPG key should be somehow included?

## Workaround

You can manually install the GPG Key file with the following puppet code:

```puppet
# Linux Software Repository Servers
# For RedHat family: https://forge.puppet.com/puppet/yum
class profile::linux::software (
  Boolean $purge = true,
  Hash $rpm_gpg_keys = {},
) {

  case $facts['os']['family'] {
    'RedHat': {
      include ::yum
      # Purge?
      resources { 'yumrepo':
        purge => $purge,
      }
      # GPG Keys
      create_resources('yum::gpgkey', $rpm_gpg_keys)
      # Resource ordering
      Yumrepo <| |> -> Yum::Gpgkey <| |> -> Package <| provider != 'rpm' |>
    }
    default: { notify { "profile::linux::software does not support osfamily: ${facts['os']['family']}": } }
  }
}
```

... and the following yaml:

```yaml
yum::managed_repos:
  - epel

profile::linux::software::rpm_gpg_keys:
  /etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7:
    content: |
      -----BEGIN PGP PUBLIC KEY BLOCK-----
      Version: GnuPG v1.4.11 (GNU/Linux)

      mQINBFKuaIQBEAC1UphXwMqCAarPUH/ZsOFslabeTVO2pDk5YnO96f+rgZB7xArB
      OSeQk7B90iqSJ85/c72OAn4OXYvT63gfCeXpJs5M7emXkPsNQWWSju99lW+AqSNm
      jYWhmRlLRGl0OO7gIwj776dIXvcMNFlzSPj00N2xAqjMbjlnV2n2abAE5gq6VpqP
      vFXVyfrVa/ualogDVmf6h2t4Rdpifq8qTHsHFU3xpCz+T6/dGWKGQ42ZQfTaLnDM
      jToAsmY0AyevkIbX6iZVtzGvanYpPcWW4X0RDPcpqfFNZk643xI4lsZ+Y2Er9Yu5
      S/8x0ly+tmmIokaE0wwbdUu740YTZjCesroYWiRg5zuQ2xfKxJoV5E+Eh+tYwGDJ
      n6HfWhRgnudRRwvuJ45ztYVtKulKw8QQpd2STWrcQQDJaRWmnMooX/PATTjCBExB
      9dkz38Druvk7IkHMtsIqlkAOQMdsX1d3Tov6BE2XDjIG0zFxLduJGbVwc/6rIc95
      T055j36Ez0HrjxdpTGOOHxRqMK5m9flFbaxxtDnS7w77WqzW7HjFrD0VeTx2vnjj
      GqchHEQpfDpFOzb8LTFhgYidyRNUflQY35WLOzLNV+pV3eQ3Jg11UFwelSNLqfQf
      uFRGc+zcwkNjHh5yPvm9odR1BIfqJ6sKGPGbtPNXo7ERMRypWyRz0zi0twARAQAB
      tChGZWRvcmEgRVBFTCAoNykgPGVwZWxAZmVkb3JhcHJvamVjdC5vcmc+iQI4BBMB
      AgAiBQJSrmiEAhsPBgsJCAcDAgYVCAIJCgsEFgIDAQIeAQIXgAAKCRBqL66iNSxk
      5cfGD/4spqpsTjtDM7qpytKLHKruZtvuWiqt5RfvT9ww9GUUFMZ4ZZGX4nUXg49q
      ixDLayWR8ddG/s5kyOi3C0uX/6inzaYyRg+Bh70brqKUK14F1BrrPi29eaKfG+Gu
      MFtXdBG2a7OtPmw3yuKmq9Epv6B0mP6E5KSdvSRSqJWtGcA6wRS/wDzXJENHp5re
      9Ism3CYydpy0GLRA5wo4fPB5uLdUhLEUDvh2KK//fMjja3o0L+SNz8N0aDZyn5Ax
      CU9RB3EHcTecFgoy5umRj99BZrebR1NO+4gBrivIfdvD4fJNfNBHXwhSH9ACGCNv
      HnXVjHQF9iHWApKkRIeh8Fr2n5dtfJEF7SEX8GbX7FbsWo29kXMrVgNqHNyDnfAB
      VoPubgQdtJZJkVZAkaHrMu8AytwT62Q4eNqmJI1aWbZQNI5jWYqc6RKuCK6/F99q
      thFT9gJO17+yRuL6Uv2/vgzVR1RGdwVLKwlUjGPAjYflpCQwWMAASxiv9uPyYPHc
      ErSrbRG0wjIfAR3vus1OSOx3xZHZpXFfmQTsDP7zVROLzV98R3JwFAxJ4/xqeON4
      vCPFU6OsT3lWQ8w7il5ohY95wmujfr6lk89kEzJdOTzcn7DBbUru33CQMGKZ3Evt
      RjsC7FDbL017qxS+ZVA/HGkyfiu4cpgV8VUnbql5eAZ+1Ll6Dw==
      =hdPa
      -----END PGP PUBLIC KEY BLOCK-----
```

### PR DESCRIPTION
**Pull request by TJM**
#### Pull Request (PR) description

Add the EPEL GPG key for EL6 and EL7, and adding the logic to automatically include GPG Keys from the yum::gpgkeys hash, including tests and some basic docs.

#### This Pull Request (PR) fixes the following issues
Fixes #95

### CODE REVIEW

File: data/os/RedHat.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+---
+## osfamily: RedHat
+
+yum::gpgkeys:
+  /etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-6: # https://archive.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-6
```
Comment (traylenator) [inline data/os/RedHat.yaml]: This will install the EPEL gpg key on osfamily RedHat even if epel is not enabled ? That seems undesirable.

Comment (TJM) [reply data/os/RedHat.yaml]: Actually, that is just the key in hieradata. It actually works similar to the "repos" and "managed_repos" logic. In order for a gpg key to be added to the system, a *repo* specifically referencing that key must be included in the `managed_repos` list. It does not simply "create_resources()" for each gpgkey in gpgkeys.  So, this specific key will not be added to the system, unless you have `epel` in manged_repos on an EL6 system (ceteris paribus).

**Comment by traylenator** [comment]
While the merge request is fine technically do we really want to maintain GPG keys here?
Maybe just using a URL
```
gpgkey => https://archive.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-7
```

**Comment by TJM** [comment]
@traylenator,

The gpgkey URL would be OK, for Internet connected machines, but in my case I need the *content* in hieradata. The only reason that we get away with specifying the file alone with the other repos is that the os-release package creates those files with the keys.

However, you are the first person who has responded in nearly a month... so, if that is all that is holding up this request, then lets change the module-level hieradata to point to the URL and I can put the "content" in my own heiradata so we can get this pulled :)

~tommy

**Comment by bastelfreak** [comment]
I think this is fine since we can provide source and/or content. Since nobody else mentioned objections I'm going to merge this.

**Pull request merged**